File: android/src/com/mygdx/game/voxel/reader/VoxelData.java
package com.mygdx.game.voxel.reader;

import android.util.SparseArray;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.HashMap;

/**
 * Created by daely on 11/24/2016.
 */

public class VoxelData {
    public final int nvoxels;
    public final int sizex,sizey,sizez;
    public ArrayList<VoxelDataPerColor> dataPerColor;
    public VoxelData(int sizex, int sizey, int sizez, MagicaVoxelDataBuffer buffer, int[] colors) {
        this.nvoxels=buffer.nvoxels;
        this.sizex=sizex;
        this.sizey=sizey;
        this.sizez=sizez;
        VoxelDataPerColor[] cmp = new VoxelDataPerColor[256];
        dataPerColor= new ArrayList<>();
        for (int icolor = 0; icolor < 256; icolor++) {
            VoxelDataPerColor vdpc=null;
            int nv=buffer.colorcount[icolor]; //n.voxel with this color
            if(nv>0) {
                vdpc=new VoxelDataPerColor(nv,colors[icolor]);
                dataPerColor.add(vdpc);
            }
            cmp[icolor]=vdpc;
        }
        //now fill it
        for (int i = 0; i < nvoxels; i++) {
            int c=buffer.color[i];
            cmp[c].addVoxel(buffer.x[i],buffer.y[i],buffer.z[i]);
        }
    }
    public int getNColors() { return dataPerColor.size(); }
}

File: android/src/com/mygdx/game/voxel/reader/VoxelData.kt
package com.mygdx.game.voxel.reader

import android.util.SparseArray

import java.lang.reflect.Array
import java.util.ArrayList
import java.util.HashMap

/**
 * Created by daely on 11/24/2016.
 */

class VoxelData internal constructor(val sizex: Int, val sizey: Int, val sizez: Int, buffer: MagicaVoxelDataBuffer, colors: IntArray) {
    val nvoxels: Int
    var dataPerColor: ArrayList<VoxelDataPerColor>

    init {
        this.nvoxels = buffer.nvoxels
        val cmp = arrayOfNulls<VoxelDataPerColor>(256)
        dataPerColor = ArrayList<VoxelDataPerColor>()
        for (icolor in 0..255) {
            var vdpc: VoxelDataPerColor? = null
            val nv = buffer.colorcount[icolor].toInt() //n.voxel with this color
            if (nv > 0) {
                vdpc = VoxelDataPerColor(nv, icolor,colors[icolor])
                dataPerColor.add(vdpc)
            }
            cmp[icolor] = vdpc
        }
        //now fill it
        for (i in 0..nvoxels - 1) {
            val c = buffer.color[i].toInt()
            cmp[c]!!.addVoxel(buffer.x[i], buffer.y[i], buffer.z[i])
        }
    }

    val nColors: Int
        get() = dataPerColor.size
}
